In [1]:
import sys
sys.path.insert(0,'/home/user/git/revitalize/neuronunit')
del sys.path[-5]
import pandas as pd
from neuronunit.optimization.optimization_management import public_opt
from sciunit.scores import RelativeDifferenceScore, ZScore
from sciunit.utils import config_set, config_get
config_set('PREVALIDATE', False)
from neuronunit.allenapi.allen_data_driven import make_allen_hard_coded_limited, make_allen_hard_coded_complete
import warnings
from neuronunit.optimization.model_parameters import BPO_PARAMS
warnings.filterwarnings("ignore")
_,_,_,a_cells = make_allen_hard_coded_limited()


[
    allen_suite_623960880,
    allen_suite_623893177,
    allen_suite471819401,
    allen_suite482493761,
    a_cells2,
    df
] = make_allen_hard_coded_complete()



# In this notebook
we fit Izhikevich and Adaptive Exponential Integrate and fire models to electrophysiology observations 
from the Allen Brain, using BluePyOpt and NeuronUnit.

# Immediately Below 
* is a list of some of the observations we will use:

In [2]:
df

TimeConstantTest       name RestingPotentialTest  \
623960880          23.8 ms  623960880             -65.1 mV   
623893177          27.8 ms  623893177             -77.0 mV   
471819401          13.8 ms  471819401             -77.5 mV   
482493761          24.4 ms  482493761             -71.6 mV   

          InputResistanceTest RheobaseTest  
623960880       241.0 megaohm      70.0 pA  
623893177       136.0 megaohm     190.0 pA  
471819401       132.0 megaohm     190.0 pA  
482493761       132.0 megaohm      70.0 pA

Allen Specimen ID: 471819401
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model AdExp

In [3]:
"""
opt,obs_preds,df = public_opt(
    a_cells['471819401'],
    BPO_PARAMS,
    '471819401',
    "ADEXP",
    100,
    200,
    "IBEA",
    score_type=RelativeDifferenceScore
    )
"""

'\nopt,obs_preds,df = public_opt(\n    a_cells[\'471819401\'],\n    BPO_PARAMS,\n    \'471819401\',\n    "ADEXP",\n    100,\n    200,\n    "IBEA",\n    score_type=RelativeDifferenceScore\n    )\n'

In [4]:
df

TimeConstantTest       name RestingPotentialTest  \
623960880          23.8 ms  623960880             -65.1 mV   
623893177          27.8 ms  623893177             -77.0 mV   
471819401          13.8 ms  471819401             -77.5 mV   
482493761          24.4 ms  482493761             -71.6 mV   

          InputResistanceTest RheobaseTest  
623960880       241.0 megaohm      70.0 pA  
623893177       136.0 megaohm     190.0 pA  
471819401       132.0 megaohm     190.0 pA  
482493761       132.0 megaohm      70.0 pA

# Solution Model Parameters:


In [6]:
#opt.attrs

Allen Specimen ID: 471819401
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model Izhikevich

In [ ]:
opt,obs_preds,df = public_opt(
    a_cells['471819401'],
    BPO_PARAMS,
    'fi_curve',
    "IZHI",
    100,
    360,
    "IBEA",
    )

RheobaseTest
TimeConstantTest
RestingPotentialTest
InputResistanceTest


INFO:__main__:gen	nevals	avg    	std    	min   	max    
1  	100   	2021.55	5233.26	155.64	53005.5
gen	nevals	avg    	std    	min   	max    
1  	100   	2021.55	5233.26	155.64	53005.5


INFO:__main__:2  	77    	1706.47	3776.46	123.486	53005.5
2  	77    	1706.47	3776.46	123.486	53005.5
INFO:__main__:3  	27    	470.277	785.894	123.486	4000   
3  	27    	470.277	785.894	123.486	4000   
INFO:__main__:4  	39    	367.732	704.635	98.6366	4000   
4  	39    	367.732	704.635	98.6366	4000   
INFO:__main__:5  	30    	317.461	605.577	98.6366	4000   
5  	30    	317.461	605.577	98.6366	4000   
INFO:__main__:6  	29    	327.117	702.073	98.6366	4000   
6  	29    	327.117	702.073	98.6366	4000   
INFO:__main__:7  	35    	337.21 	621.931	98.6366	4000   
7  	35    	337.21 	621.931	98.6366	4000   
INFO:__main__:8  	29    	263.531	467.082	98.6366	4000   
8  	29    	263.531	467.082	98.6366	4000   
INFO:__main__:9  	34    	190.609	284.276	98.6365	1108.13
9  	34    	190.609	284.276	98.6365	1108.13
INFO:__main__:10 	33    	281.454	497.015	98.6365	4000   
10 	33    	281.454	497.015	98.6365	4000   
INFO:__main__:11 	32    	271.855	593.204	98.6365	4000   
11 	32    	271.855	593.204	98.6365	4000   


INFO:__main__:84 	35    	308.118	387.541	50.0449	1065.72
84 	35    	308.118	387.541	50.0449	1065.72
INFO:__main__:85 	32    	242.755	362.318	50.0449	1054.31
85 	32    	242.755	362.318	50.0449	1054.31
INFO:__main__:86 	26    	272.678	371.763	50.0449	1098.42
86 	26    	272.678	371.763	50.0449	1098.42
INFO:__main__:87 	33    	329.723	382.409	50.0449	1070.29
87 	33    	329.723	382.409	50.0449	1070.29
INFO:__main__:88 	24    	275.916	354.832	50.0449	1050.52
88 	24    	275.916	354.832	50.0449	1050.52
INFO:__main__:89 	32    	279.186	380.668	50.0449	1059.69
89 	32    	279.186	380.668	50.0449	1059.69
INFO:__main__:90 	33    	299.996	381.694	50.0449	1182.41
90 	33    	299.996	381.694	50.0449	1182.41
INFO:__main__:91 	36    	241.07 	364.727	50.0449	1065.62
91 	36    	241.07 	364.727	50.0449	1065.62
INFO:__main__:92 	34    	308.879	388.07 	50.0449	1075.22
92 	34    	308.879	388.07 	50.0449	1075.22
INFO:__main__:93 	29    	305.123	400.907	50.0449	1053.89
93 	29    	305.123	400.907	50.0449	1053.89


INFO:__main__:166	40    	329.412	409.085	23.1867	1270.2 
166	40    	329.412	409.085	23.1867	1270.2 
INFO:__main__:167	29    	306.757	417.574	23.1867	1054.13
167	29    	306.757	417.574	23.1867	1054.13
INFO:__main__:168	27    	224.987	315.714	23.1867	1043.31
168	27    	224.987	315.714	23.1867	1043.31
INFO:__main__:169	35    	305.959	384.644	23.1867	1046.06
169	35    	305.959	384.644	23.1867	1046.06
INFO:__main__:170	37    	335.442	416.213	23.1867	1055.14
170	37    	335.442	416.213	23.1867	1055.14
INFO:__main__:171	38    	316.353	395.14 	22.0994	1074.62
171	38    	316.353	395.14 	22.0994	1074.62
INFO:__main__:172	24    	261.666	320.098	22.0994	1040.63
172	24    	261.666	320.098	22.0994	1040.63
INFO:__main__:173	40    	387.109	409.022	22.0994	1077.12
173	40    	387.109	409.022	22.0994	1077.12
INFO:__main__:174	33    	367.992	409.438	22.0994	1064.58
174	33    	367.992	409.438	22.0994	1064.58
INFO:__main__:175	21    	233.787	339.264	22.0994	1043.63
175	21    	233.787	339.264	22.0994	1043.63


INFO:__main__:248	32    	314.312	411.642	19.7813	1058.45
248	32    	314.312	411.642	19.7813	1058.45
INFO:__main__:249	38    	282.324	417.593	19.7813	1073.57
249	38    	282.324	417.593	19.7813	1073.57
INFO:__main__:250	37    	261.208	383.303	19.7813	1047.86
250	37    	261.208	383.303	19.7813	1047.86
INFO:__main__:251	28    	231.084	377.323	19.7813	1062.76
251	28    	231.084	377.323	19.7813	1062.76
INFO:__main__:252	34    	281.154	398.391	19.7813	1097.71
252	34    	281.154	398.391	19.7813	1097.71
INFO:__main__:253	32    	230.311	383.947	19.7813	1067.21
253	32    	230.311	383.947	19.7813	1067.21
INFO:__main__:254	28    	225.906	350.988	19.7813	1050   
254	28    	225.906	350.988	19.7813	1050   
INFO:__main__:255	39    	274.064	399.549	19.7813	1040.47
255	39    	274.064	399.549	19.7813	1040.47
INFO:__main__:256	25    	246.588	357.269	19.7813	1076.14
256	25    	246.588	357.269	19.7813	1076.14
INFO:__main__:257	39    	256.315	394.01 	19.7813	1074.94
257	39    	256.315	394.01 	19.7813	1074.94


In [ ]:
df

# Solution Parameters:

In [ ]:
opt.attrs

# Just FI curve on AdExp model
Allen Specimin Id: 471819401

In [ ]:
opt,obs_preds,df = public_opt(
    a_cells['fi_curve'],
    BPO_PARAMS,
    'fi_curve',
    "ADEXP",
    100,
    100,
    "IBEA"
    )

In [ ]:
df


In [ ]:
opt.attrs

Allen Specimen ID: 623960880
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model Izhikevich

In [ ]:

opt,obs_preds,df = public_opt(
    allen_suite_623960880.tests,
    BPO_PARAMS,
    allen_suite_623960880.name,
    "IZHI",
    100,
    100,
    "IBEA"
    )

In [ ]:
df

In [ ]:
opt.attrs